kernel/byterun/coq_interp.c

```diff
@@ -1697,6 +1697,30 @@ value coq_interprete
         Next;
       }
 
+      Instruct (CHECKEQUALFLOAT) {
+        double x, y;
+        print_instr("CHECKEQUALFLOAT");
+        CheckFloat2();
+        x = Double_val(accu);
+        y = Double_val(*sp++);
+        switch (fpclassify(x)) {
+        case FP_NORMAL:
+        case FP_SUBNORMAL:
+        case FP_INFINITE:
+          accu = (x == y) ? coq_true : coq_false;
+          break;
+        case FP_ZERO:
+          accu = (x == y && !signbit(x) == !signbit(y)) ? coq_true : coq_false;
+          /* note that the spec of signbit only says "non negative value"
+             so we need to normalize booleans with ! before comparing them */
```

Comment:
I suppose you meant "nonzero".

Regarding this whole code, it might be worth delaying the classification to improve efficiency, something like:
```c
if (x == y) {
  accu = (x != 0 || !signbit(x) == !signbit(y)) ? coq_true : coq_false;
} else {
  accu = (x != x && y != y) ? coq_true : coq_false;
}
```

Reply:
Indeed, did the same for the OCaml implem.